go/libraries/doltcore/env/actions/checkout.go

```diff
@@ -92,3 +97,539 @@ func MoveTablesFromHeadToWorking(ctx context.Context, roots doltdb.Roots, tbls [
+			if err != nil {
+				return nil, err
+			}
+
+			if oldHash == emptyHash {
+				resultMap[tblName] = changedHash
+			} else if force {
+				resultMap[tblName] = oldHash
+			} else if oldHash != changedHash {
+				conflicts.Add(tblName)
+			}
+		}
+	}
+
+	return resultMap, nil
+}
+
+// moveForeignKeys returns the foreign key collection that should be used for the new working set.
+func moveForeignKeys(ctx context.Context, oldRoot, newRoot, changedRoot *doltdb.RootValue, force bool) (*doltdb.ForeignKeyCollection, error) {
+	oldFks, err := oldRoot.GetForeignKeyCollection(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	newFks, err := newRoot.GetForeignKeyCollection(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	changedFks, err := changedRoot.GetForeignKeyCollection(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	oldHash, err := oldFks.HashOf(ctx, oldRoot.VRW())
+	if err != nil {
+		return nil, err
+	}
+
+	newHash, err := newFks.HashOf(ctx, newRoot.VRW())
+	if err != nil {
+		return nil, err
+	}
+
+	changedHash, err := changedFks.HashOf(ctx, changedRoot.VRW())
+	if err != nil {
+		return nil, err
+	}
+
+	if oldHash == changedHash {
+		return newFks, nil
+	} else if oldHash == newHash {
+		return changedFks, nil
+	} else if newHash == changedHash {
+		return oldFks, nil
+	} else {
+		// Both roots have modified the foreign keys. We need to do more work to merge them together into a new foreign
+		// key collection.
+		return mergeForeignKeyChanges(ctx, oldFks, newRoot, newFks, changedRoot, changedFks, force)
+	}
+}
+
+// mergeForeignKeyChanges merges the foreign key changes from the old and changed roots into a new foreign key
+// collection, or returns an error if the changes are incompatible. Changes are incompatible if the changed root
+// and new root both altered foreign keys on the same table.
+func mergeForeignKeyChanges(
+	ctx context.Context,
+	oldFks *doltdb.ForeignKeyCollection,
+	newRoot *doltdb.RootValue,
+	newFks *doltdb.ForeignKeyCollection,
+	changedRoot *doltdb.RootValue,
+	changedFks *doltdb.ForeignKeyCollection,
+	force bool,
+) (*doltdb.ForeignKeyCollection, error) {
+	fksByTable := make(map[string][]doltdb.ForeignKey)
+
+	conflicts := set.NewEmptyStrSet()
+	tblNames, err := newRoot.GetTableNames(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, tblName := range tblNames {
+		oldFksForTable, _ := oldFks.KeysForTable(tblName)
+		newFksForTable, _ := newFks.KeysForTable(tblName)
+		changedFksForTable, _ := changedFks.KeysForTable(tblName)
+
+		oldHash := doltdb.CombinedHash(oldFksForTable)
+		newHash := doltdb.CombinedHash(newFksForTable)
+		changedHash := doltdb.CombinedHash(changedFksForTable)
+
+		if oldHash == changedHash {
+			fksByTable[tblName] = append(fksByTable[tblName], newFksForTable...)
+		} else if oldHash == newHash {
+			fksByTable[tblName] = append(fksByTable[tblName], changedFksForTable...)
+		} else if newHash == changedHash {
+			fksByTable[tblName] = append(fksByTable[tblName], oldFksForTable...)
```

Comment:
Similar comment here... seems like if the working/staged set has the same FKs as the new root we're moving to (and they differ from the current HEAD root), then we'd want to return new/changed FKs.

I don't think the BATS tests would trigger this case where you move to a branch that has the same FKs as were added in the working set, so maybe adding that will confirm if this is the right code or not.

Reply:
Same deal.